<a href="https://colab.research.google.com/github/AshikSathiya/Prediction-of-Product-Sales/blob/main/Prediction_of_Product_Sales_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# Import pandas as change max columns
import pandas as pd
import numpy as np


from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder

from sklearn.model_selection import train_test_split
from sklearn.compose import make_column_selector, ColumnTransformer

from sklearn import set_config
set_config(transform_output='pandas')

In [3]:
fpath = '/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv'
df = pd.read_csv(fpath)
df

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8523 rows x 12 columns]

#Cleaning Process

##Drop Duplicates

In [4]:
# Duplicate Filter

duplicated_rows = df.duplicated()
duplicated_rows

# Calculate the sum of the duplicated_rows filter
duplicated_rows.sum()

0

There are no duplicates

##Inconsitencies

In [5]:
object_cols = df.select_dtypes('object').columns


In [6]:
# Check the nunique for just the object cols
object_nunique = df[object_cols].nunique()
object_nunique

Item_Identifier         1559
Item_Fat_Content           5
Item_Type                 16
Outlet_Identifier         10
Outlet_Size                3
Outlet_Location_Type       3
Outlet_Type                4
dtype: int64

It appears as though Item_Fat_Content, Outlet_Size, Outlet_Location_Type, and Outlet_Type have a small amount of values, we will look more into these columns

In [7]:
# Drop the columns listed above
df.drop(["Item_Identifier", "Item_Type", "Outlet_Identifier"], axis=1, inplace=True)
df.head()

Item_Weight Item_Fat_Content  Item_Visibility  Item_MRP  \
0         9.30          Low Fat         0.016047  249.8092   
1         5.92          Regular         0.019278   48.2692   
2        17.50          Low Fat         0.016760  141.6180   
3        19.20          Regular         0.000000  182.0950   
4         8.93          Low Fat         0.000000   53.8614   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [8]:
# Remake the list of string columns (after dropping previous cols)
# loop through the list of string columns and print the value counts for the column


string_columns = df.select_dtypes(include='object').columns

for col in string_columns:
    value_counts = df[col].value_counts()
    print(value_counts)
    print()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: Outlet_Type, dtype: int64



Inconsitent values are present in the Item_Fat_Content

In [9]:
# Check for inconsistencies by inspecting the value_counts for Item_Fat_Content
Item_Fat_Content_counts = df['Item_Fat_Content'].value_counts()
print(Item_Fat_Content_counts)

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64



Rather than having 5 values, we should only have 2 (Low Fat and Regular)

In [10]:
# Standardize the values in the Central column
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({"LF":"Low Fat"})
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({"low fat":"Low Fat"})
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({"reg":"Regular"})


df['Item_Fat_Content'].value_counts()
# Check the value counts again to confirm
Item_Fat_Content_counts = df['Item_Fat_Content'].value_counts()
print(Item_Fat_Content_counts)

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64


##Split Data

In [11]:
y = df['Item_Outlet_Sales']
X = df.drop('Item_Outlet_Sales', axis = 1)

display(y)

display(X)

0       3735.1380
1        443.4228
2       2097.2700
3        732.3800
4        994.7052
          ...    
8518    2778.3834
8519     549.2850
8520    1193.1136
8521    1845.5976
8522     765.6700
Name: Item_Outlet_Sales, Length: 8523, dtype: float64

Item_Weight Item_Fat_Content  Item_Visibility  Item_MRP  \
0           9.300          Low Fat         0.016047  249.8092   
1           5.920          Regular         0.019278   48.2692   
2          17.500          Low Fat         0.016760  141.6180   
3          19.200          Regular         0.000000  182.0950   
4           8.930          Low Fat         0.000000   53.8614   
...           ...              ...              ...       ...   
8518        6.865          Low Fat         0.056783  214.5218   
8519        8.380          Regular         0.046982  108.1570   
8520       10.600          Low Fat         0.035186   85.1224   
8521        7.210          Regular         0.145221  103.1332   
8522       14.800          Low Fat         0.044878   75.4670   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  
0     Supermarket Type1  
1     Supermarket Type2  
2     Supermarket Type1  
3         Grocery Store  
4     Supermarket Type1  
...                 ...  
8518  Supermarket Type1  
8519  Supermarket Type1  
8520  Supermarket Type1  
8521  Supermarket Type2  
8522  Supermarket Type1  

[8523 rows x 8 columns]

In [12]:
# Train test split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)
X_train.info()
y_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6392 entries, 4776 to 7270
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Weight                5285 non-null   float64
 1   Item_Fat_Content           6392 non-null   object 
 2   Item_Visibility            6392 non-null   float64
 3   Item_MRP                   6392 non-null   float64
 4   Outlet_Establishment_Year  6392 non-null   int64  
 5   Outlet_Size                4580 non-null   object 
 6   Outlet_Location_Type       6392 non-null   object 
 7   Outlet_Type                6392 non-null   object 
dtypes: float64(3), int64(1), object(4)
memory usage: 449.4+ KB
<class 'pandas.core.series.Series'>
Int64Index: 6392 entries, 4776 to 7270
Series name: Item_Outlet_Sales
Non-Null Count  Dtype  
--------------  -----  
6392 non-null   float64
dtypes: float64(1)
memory usage: 99.9 KB


##Create a preprocessing object to prepare the dataset for Machine Learning

In [13]:
X_train.select_dtypes('object')

Item_Fat_Content Outlet_Size Outlet_Location_Type        Outlet_Type
4776          Low Fat      Medium               Tier 3  Supermarket Type2
7510          Regular      Medium               Tier 3  Supermarket Type2
5828          Regular      Medium               Tier 1  Supermarket Type1
5327          Low Fat       Small               Tier 2  Supermarket Type1
4810          Low Fat         NaN               Tier 2  Supermarket Type1
...               ...         ...                  ...                ...
5734          Regular         NaN               Tier 3      Grocery Store
5191          Low Fat         NaN               Tier 2  Supermarket Type1
5390          Low Fat         NaN               Tier 2  Supermarket Type1
860           Low Fat         NaN               Tier 2  Supermarket Type1
7270          Low Fat       Small               Tier 1  Supermarket Type1

[6392 rows x 4 columns]

It appears that Outlet Size snd Outlet Location Type are our ordinal values

###Ordinal Pipeline

In [14]:
df['Outlet_Size'].nunique()


3

In [15]:
ord_col = ['Outlet_Size','Outlet_Location_Type']

# Impute Most frequent Strategy
impute_common = SimpleImputer(strategy='most_frequent')

# SPecify Order of Ordinal COLUMNS
size_order = ['Small','Medium','High']
loc_order= ['Tier 1','Tier 2','Tier 3']
ord_encoder = OrdinalEncoder(categories=[size_order, loc_order])

scaler = StandardScaler()

# make a pipeline
ord_pipe = make_pipeline(impute_common, ord_encoder, scaler)
ord_pipe


Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='most_frequent')),
                ('ordinalencoder',
                 OrdinalEncoder(categories=[['Small', 'Medium', 'High'],
                                            ['Tier 1', 'Tier 2', 'Tier 3']])),
                ('standardscaler', StandardScaler())])

###Categorical Pipeline

In [16]:
# Prepare Categorical pipeline
cat_cols = X_train.select_dtypes('object').drop(columns=ord_col).columns

#Impute Missing values with "MISSING"
impute_missing = SimpleImputer(strategy='constant',fill_value='MISSING')

#One Hot Encoding
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

#Create pipline
ohe_pipe = make_pipeline(impute_missing, ohe_encoder)
ohe_pipe

Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='MISSING', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

###Numeric Pipeline

In [17]:
#Prepare Numeric piepline
num_cols = X_train.select_dtypes('number').columns

# Impute using mean
mean_imputer = SimpleImputer(strategy="mean")

#Scaling
scaler = StandardScaler()

#Numeric Pipeline
numeric_pipe = make_pipeline(mean_imputer, scaler)
numeric_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler())])

##Create a Column Transformer

###Create Tuples

In [18]:
# Define three tuples

# Making a numeric tuple for ColumnTransformer
num_tuple = ('numeric', numeric_pipe, num_cols)

# Making a ohe_tuple for ColumnTransformer
ohe_tuple = ('categorical', ohe_pipe, cat_cols)


# Making an ordinal_tuple for ColumnTransformer
ord_tuple = ('ordinal', ord_pipe, ord_col)


###Create Column Transformer

In [19]:
col_transformer = ColumnTransformer([num_tuple, ord_tuple, ohe_tuple],
                                    verbose_feature_names_out=False)

col_transformer

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  OrdinalEncoder(categ...
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['Outlet_Size', 'Outlet_Location_Type']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Outlet_Type'], dtype='object'))],
                  verbose_feature_names_out=False)

###Transform

In [20]:
#Fit the column transformer object to the training data.
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('ordinal',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('ordinalencoder',
                                                  OrdinalEncoder(categ...
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 ['Outlet_Size', 'Outlet_Location_Type']),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Outlet_Type'], dtype='object'))],
                  verbose_feature_names_out=False)

In [21]:
# Transform the training data
X_train_processed = col_transformer.transform(X_train)
X_train_processed.head()


Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.817249        -0.712775  1.828109                   1.327849   
7510     0.556340        -1.291052  0.603369                   1.327849   
5828    -0.131512         1.813319  0.244541                   0.136187   
5327    -1.169219        -1.004931 -0.952591                   0.732018   
4810     1.528819        -0.965484 -0.336460                   0.493686   

      Outlet_Size  Outlet_Location_Type  Item_Fat_Content_Low Fat  \
4776     0.287374              1.084948                       1.0   
7510     0.287374              1.084948                       0.0   
5828     0.287374             -1.384777                       0.0   
5327    -1.384048             -0.149914                       1.0   
4810     0.287374             -0.149914                       1.0   

      Item_Fat_Content_Regular  Outlet_Type_Grocery Store  \
4776                       0.0                        0.0   
7510                       1.0                        0.0   
5828                       1.0                        0.0   
5327                       0.0                        0.0   
4810                       0.0                        0.0   

      Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
4776                            0.0                            1.0   
7510                            0.0                            1.0   
5828                            1.0                            0.0   
5327                            1.0                            0.0   
4810                            1.0                            0.0   

      Outlet_Type_Supermarket Type3  
4776                            0.0  
7510                            0.0  
5828                            0.0  
5327                            0.0  
4810                            0.0

In [22]:
# Transform the testing data
X_test_processed = col_transformer.transform(X_test)
X_test_processed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
7503  3.310089e-01        -0.776646 -0.998816                  -1.293807   
2957 -1.179892e+00         0.100317 -1.585194                  -0.102145   
7031  3.784469e-01        -0.482994 -1.595784                   0.136187   
1084  4.213344e-16        -0.415440  0.506592                  -1.532139   
856  -6.426567e-01        -1.047426  0.886725                   0.732018   

      Outlet_Size  Outlet_Location_Type  Item_Fat_Content_Low Fat  \
7503     1.958796              1.084948                       1.0   
2957    -1.384048             -1.384777                       1.0   
7031     0.287374             -1.384777                       0.0   
1084     0.287374              1.084948                       0.0   
856     -1.384048             -0.149914                       0.0   

      Item_Fat_Content_Regular  Outlet_Type_Grocery Store  \
7503                       0.0                        0.0   
2957                       0.0                        0.0   
7031                       1.0                        0.0   
1084                       1.0                        0.0   
856                        1.0                        0.0   

      Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
7503                            1.0                            0.0   
2957                            1.0                            0.0   
7031                            1.0                            0.0   
1084                            0.0                            0.0   
856                             1.0                            0.0   

      Outlet_Type_Supermarket Type3  
7503                            0.0  
2957                            0.0  
7031                            0.0  
1084                            1.0  
856                             0.0

#CRISP-DM Phase 4 - Modeling



##Imports

In [23]:
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, GridSearchCV


In [24]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = mean_squared_error(y_true, y_pred, squared=False)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics
def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)
  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  print()
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )
  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set the label as the index
    results_df = results_df.set_index('Label')
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)

##1. Your first task is to build a linear regression model to predict sales.



###Build a linear regression model.


In [25]:
# Instantiate linear regression model
linreg = LinearRegression()


# Combine your preprocessing object (from the previous step of the project) and the model into a model pipeline.
linreg_pipe = make_pipeline(col_transformer, linreg)
linreg_pipe

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer()),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  (...
                                                  ['Outlet_Size',
                                                   'Outlet_Location_Type']),
                                                 ('categorical',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='MISSING',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Outlet_Type'], dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('linearregression', LinearRegression())])

In [26]:
# Fit the training data on to model pipeline
linreg_pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer()),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  (...
                                                  ['Outlet_Size',
                                                   'Outlet_Location_Type']),
                                                 ('categorical',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='MISSING',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Outlet_Type'], dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('linearregression', LinearRegression())])

###Use the custom evaluation function to get the metrics for your model (on training and test data).

In [27]:
#Model Evaluation
evaluate_regression(linreg_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 848.394
- MSE = 1,302,670.477
- RMSE = 1,141.346
- R^2 = 0.560

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 803.991
- MSE = 1,190,707.288
- RMSE = 1,091.195
- R^2 = 0.568


###Compare the training vs. test R-squared values and answer the question: to what extent is this model overfit/underfit?






**Mean Absolute Error (MAE):** For the training data, the MAE is 848.394, and for the test data, it is 803.991. This means that on average, the model's predictions are off by approximately 848 units for the training data and 804 units for the test data.


**Mean Squared Error (MSE):** The MSE is 1,302,670.477 for the training data and 1,190,707.288 for the test data. Higher MSE values for the training data indicate larger prediction errors compared to the test data.

**Root Mean Squared Error (RMSE)**: RMSE is the square root of MSE and so it provides a similar measure of accuracy. The RMSE is 1,141.346 for the training data and 1,091.195 for the test data. Similar to MSE, the higher RMSE for the training data suggests larger errors compared to the testing data.

**R^2 (Coefficient of Determination)**: R^2 represents the proportion of the variance in sales that is predictable from the independent variables used in the model. The R^2 is 0.560 for the training data and 0.568 for the test data. This suggests that approximately 56% of the variance in sales can be explained by the model for the training data, and 57% for the test data.

**Overfitting/Underfitting**:

Overfitting occurs when a model performs well on the training data but poorly on test data due to the data being unseen. In this case, the model's performance on the training and test data is relatively similar across all metrics, with the test metrics actually being slightly better. This indicates that overfitting is less likely.

Underfitting occurs when a model is unable understand the underlying patterns in the data. The R^2 values of 0.560 for training data and 0.568 for test data suggest that the model may be somewhat underfitting and indicates that the model can not predict a substantial portion of the variance in sales.



##2. Your second task is to build a Random Forest model to predict sales.



###Build a default Random Forest model.


In [28]:
# Instantiate a Default Model
rf = RandomForestRegressor(random_state = 42)

In [29]:
# Create pipeline for default model
rf_pipe = make_pipeline(col_transformer, rf)

In [30]:
# Fit the default model on the training data
rf_pipe.fit(X_train, y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer()),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('ordinal',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  (...
                                                   'Outlet_Location_Type']),
                                                 ('categorical',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='MISSING',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Outlet_Type'], dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('randomforestregressor',
                 RandomForestRegressor(random_state=42))])

###Use the custom evaluation function to get the metrics for your model (on training and test data).

In [31]:
# Use custom function to predict and evaluate
evaluate_regression(rf_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 298.133
- MSE = 184,157.037
- RMSE = 429.135
- R^2 = 0.938

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 774.841
- MSE = 1,235,256.823
- RMSE = 1,111.421
- R^2 = 0.552


###Compare the training vs. test R-squared values and answer the question: to what extent is this model overfit/underfit?



The R-squared value of 0.938 from the training data indicates that approximately 93.8% of the variance in the target variable can be explained by the independent variables in the model. This is a reasonably good fit to the training data.

However the testing data metrics were considerably poorer. The R-squared value of 0.552 for the test data indicates that approximately 55.2% of the variance in the sales can be explained by the model. This value is lower compared to the training R-squared, suggesting overfitting

Overfitting occurs when a model performs significantly better on the training data compared to the test data, indicating that it may have memorized the training data's patterns instead of learning the underlying general patterns.





###Compare this model's performance to the linear regression model: which model has the best test scores?


When comparing the metrics from the random forest model to the linear regression model. It is evident that the linear regression model performed better. The linear regression model had a higher R^2 of 0.568 and the random forest model performed slightly worse with a score of 0.552.

##3. Use GridSearchCV to tune at least two hyperparameters for a Random Forest model.



In [32]:
# Parameters for tuning
rf_pipe.get_params()

{'memory': None,
 'steps': [('columntransformer',
   ColumnTransformer(transformers=[('numeric',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer()),
                                                    ('standardscaler',
                                                     StandardScaler())]),
                                    Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
          'Outlet_Establishment_Year'],
         dtype='object')),
                                   ('ordinal',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer(strategy='most_frequent')),
                                                    ('ordinalencoder',
                                                     OrdinalEncoder(categ...
                                                    ('standardscaler',
                              

In [33]:
#Create parameter list to test

params = { 'randomforestregressor__max_features': [1.0, 2.0 , 3.0, 4.0],
           'randomforestregressor__max_depth': [None, 10 , 15, 20, 25],
           'randomforestregressor__n_estimators': [50, 100, 150, 200],
           'randomforestregressor__max_features':['sqrt','log2',None],
           'randomforestregressor__oob_score': [False,True],
          'randomforestregressor__warm_start': [False,True],
         }


In [34]:
# Instaniate the gridsearch
gridsearch = GridSearchCV(rf_pipe, params, n_jobs=-1, cv = 5, verbose=1)
gridsearch

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('columntransformer',
                                        ColumnTransformer(transformers=[('numeric',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImputer()),
                                                                                         ('standardscaler',
                                                                                          StandardScaler())]),
                                                                         Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                                        ('ordinal',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImputer(...
                                       ('randomforestregressor',
                                        RandomForestRegressor(random_state=42))]),
             n_jobs=-1,
             param_grid={'randomforestregressor__max_depth': [None, 10, 15, 20,
                                                              25],
                         'randomforestregressor__max_features': ['sqrt', 'log2',
                                                                 None],
                         'randomforestregressor__n_estimators': [50, 100, 150,
                                                                 200],
                         'randomforestregressor__oob_score': [False, True],
                         'randomforestregressor__warm_start': [False, True]},
             verbose=1)

In [35]:
# Fit the gridsearch on the training data
gridsearch.fit(X_train, y_train)

# Obtain the best paramters from the gridsearch
gridsearch.best_params_

Fitting 5 folds for each of 240 candidates, totalling 1200 fits


{'randomforestregressor__max_depth': 10,
 'randomforestregressor__max_features': None,
 'randomforestregressor__n_estimators': 150,
 'randomforestregressor__oob_score': False,
 'randomforestregressor__warm_start': False}

###After determining the best parameters from your GridSearch, fit and evaluate a final best model on the entire training set (no folds).


In [36]:
# Define and refit best model
best_rf = gridsearch.best_estimator_
evaluate_regression(best_rf, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 640.516
- MSE = 821,866.774
- RMSE = 906.569
- R^2 = 0.722

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 739.251
- MSE = 1,130,000.683
- RMSE = 1,063.015
- R^2 = 0.590


###Compare your tuned model to your default Random Forest: did the performance improve?

Overall the tuned Model of my default Random Forest model definetly improved in performance. The training set decreased in predictive performance however the test set increased in performance. This reduced overfitting, however did not completely solve the issue of overfitting as it is still present. All in all we did have a better R^2 score in our tuned model with a score of 0.590 which is greater than the score 0.552 which is the R^2 score from our default model.





#CRISP-DM Phase 5 - Evaluation



##4. You now have tried several different models on your data set. You need to determine which model to implement.



###Overall, which model do you recommend?
###Justify your recommendation.

After creating and evalaluting three models (default linear regression model, default random forest model, tuned random forest model), it is clear that the best model to use is the **tuned random forest model**.



* I have chosen the tuned random forest model because it has the highest R^2 score on the test data so it can most accuratley predict sales prices on unseen data. Along with this, it has the lowest instance of overfitting or undefitting.

* In the default random forest model, the R-squared value of 0.938 from the training data the R-squared value of 0.552 for the test data the model is heavily overfitted.


* In the default linear regression model, our model was very much underfitted. Underfitting occurs when a model is unable understand the patterns within the data and resorts in a very innacurate model. The R^2 values of 0.560 for training data and 0.568 for test data shows that the model is underfitted as it resorts to practically guessing the sales price.

#Interpret Chosen Models Performance for Stake Holder

##Understanding R-squared Value

R-squared (R^2) is a metric that helps assess the performance of a regression model. R^2 value indicates how well the model can predict the dependent variable based on the other features(independent variables). Higher R^2 values indicate a  ability to make accurate predictions, while lower values indicate less accuracy.

In our tuned random forest model,the R^2 value for the training data is 0.722, meaning that approximately 72.2% of the variability in the dependent variable is explained by the independent variables used in the model. This is a reasonably good fit between the model and the training data.

For the test data, the R^2 value is 0.590. While this is a lower value compared to the training data, it still suggests that the model is able to explain a significant portion of the relationships present in the test data.

##Understanding the MAE Metric

The MAE measures the average absolute difference between the predicted and actual values. It provides a clear understanding of the prediction errors. In our model, the MAE value for the test data is 739.251. This means that, when using the model to make predictions on unseen data, we can expect the average prediction error to be around 739.251 unit.

By providing the MAE metric, we can show the stakeholder a straightforward measure of prediction accuracy. The stakeholder can understand that the model's average prediction error on the test data is 739.251 units, which can help them assess the reliability of the model's predictions and make decisions based on the expected levels of accuracy and error.





